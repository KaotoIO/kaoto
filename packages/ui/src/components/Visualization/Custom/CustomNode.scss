/* stylelint-disable-next-line selector-class-pattern */
.custom-node__image {
  display: flex;
  flex-flow: column;
  align-items: center;
  height: 100%;
  justify-content: center;

  img {
    max-height: 80%;
  }
}
