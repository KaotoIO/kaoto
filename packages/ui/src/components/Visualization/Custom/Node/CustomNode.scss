/* stylelint-disable */

.custom-node {
  &__image {
    display: flex;
    flex-flow: column;
    align-items: center;
    height: 100%;
    justify-content: center;

    img {
      max-height: 80%;
      max-width: 95%;
    }
  }

  &[data-disabled='true'] &__image {
    filter: grayscale(100%);
  }

  &--disabled {
    --pf-topology__node_decorator--Color: var(--pf-v5-global--BackgroundColor--dark-200);
  }
}

.custom-node__label--disabled {
  font-style: italic;
  text-decoration: line-through;
}

.custom-node__label {
  z-index: 20;
}

/* Ensure selected state has higher specificity */
.custom-node__label--selected, .custom-node__label--selected:hover {
  --pf-topology__node__label__background--Stroke: var(--pf-topology__node--m-selected--node__label__background--Stroke);
  --pf-topology__node__label__background--Fill: var(--pf-topology__node--m-selected--node__label__background--Fill);
  --pf-topology__node__separator--Stroke: var(--pf-topology__node--m-selected--node__label__text--m-secondary--Fill);
  --pf-topology__node__action-icon__icon--Color: var(--pf-topology__node--m-selected--action-icon__icon--Color);
  --pf-topology__node__label__text--Fill: var(--pf-topology__node--m-selected--node__label__text--Fill);
  --pf-topology__node__label__text--m-secondary--Fill: var(--pf-topology__node--m-selected--node__label__text--m-secondary--Fill);

  z-index: 50;
}