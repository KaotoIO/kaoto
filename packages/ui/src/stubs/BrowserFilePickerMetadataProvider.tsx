import { ChangeEvent, createRef, FunctionComponent, PropsWithChildren, useCallback, useRef } from 'react';
import {
  SCHEMA_FILE_ACCEPT_PATTERN,
  SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY,
  SCHEMA_FILE_NAME_PATTERN,
  SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY,
} from '../models/datamapper';
import { IMetadataApi, MetadataContext } from '../providers';
import { readFileAsString } from './read-file-as-string';

export const BrowserFilePickerMetadataProvider: FunctionComponent<PropsWithChildren> = (props) => {
  const fileInputRef = createRef<HTMLInputElement>();
  const fileSelectionRef = useRef<{
    resolve: (files: Record<string, string>) => void;
    reject: (error: unknown) => unknown;
  }>();
  const fileContentsRef = useRef<Record<string, string>>();

  const askUserForFileSelection = useCallback(
    (
      include: string,
      _exclude?: string,
      _options?: Record<string, unknown>,
    ): Promise<string[] | string | undefined> => {
      if (!fileInputRef.current) return Promise.resolve(undefined);

      if (include === SCHEMA_FILE_NAME_PATTERN_SOURCE_BODY) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN_SOURCE_BODY;
      } else if (include === SCHEMA_FILE_NAME_PATTERN) {
        fileInputRef.current.accept = SCHEMA_FILE_ACCEPT_PATTERN;
      }

      fileInputRef.current.click();
      return new Promise<Record<string, string>>((resolve, reject) => {
        fileSelectionRef.current = { resolve, reject };
      }).then((files) => {
        fileContentsRef.current = files;
        return Object.keys(files);
      });
    },
    [fileInputRef],
  );

  const onImport = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const schemaFiles = event.target.files;
    if (!schemaFiles) return;
    const fileContents: Record<string, string> = {};
    const fileContentPromises: Promise<string>[] = [];
    Array.from(schemaFiles).forEach((f) => {
      const promise = readFileAsString(f).then((content) => (fileContents[f.name] = content));
      fileContentPromises.push(promise);
    });
    await Promise.allSettled(fileContentPromises);

    fileSelectionRef.current?.resolve(fileContents);
    fileSelectionRef.current = undefined;
    event.target.value = '';
  }, []);

  const getResourceContent = useCallback((path: string) => {
    return Promise.resolve(fileContentsRef.current && fileContentsRef.current[path]);
  }, []);

  const metadataApi: IMetadataApi = {
    askUserForFileSelection: askUserForFileSelection,
    getResourceContent: getResourceContent,
    shouldSaveSchema: true,
    getMetadata: () => Promise.resolve(undefined),
    setMetadata: () => Promise.resolve(),
    deleteResource: () => Promise.resolve(true),
    saveResourceContent: () => Promise.resolve(),
    getSuggestions: () => Promise.resolve([]),
    onStepUpdated: () => Promise.resolve(),
  };

  return (
    <MetadataContext.Provider value={metadataApi}>
      {props.children}
      <input
        type="file"
        style={{ display: 'none' }}
        data-testid={`attach-schema-file-input`}
        onChange={onImport}
        ref={fileInputRef}
      />
    </MetadataContext.Provider>
  );
};
