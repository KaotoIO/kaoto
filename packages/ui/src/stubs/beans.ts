export const beansYaml = `
- beans:
  - name: myBean
    type: io.kaoto.MyBean
    properties:
      p1: p1v
      p2: p2v
  - name: myBean2
    type: io.kaoto.MyBean
    properties:
      p1:
        p1s1: p1s1v
`

export const beansJson = {
  "beans": [
    {
      "name": "myBean",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": "p1v",
        "p2": "p2v"
      }
    },
    {
      "name": "myBean2",
      "type": "io.kaoto.MyBean",
      "properties": {
        "p1": {
          "p1s1": "p1s1v"
        }
      }
    }
  ]
}

