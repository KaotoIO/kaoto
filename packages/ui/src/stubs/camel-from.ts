import { FromDefinition } from '@kaoto-next/camel-catalog/types';
import { CamelRouteVisualEntity } from '../models/visualization/flows';

/**
 * This is a stub Camel From in YAML format.
 * It is used to test the Canvas component.
 */
export const camelFromYaml = `
- from:
    uri: timer:tutorial
    steps:
    - to:
        uri: direct:my-route
`;

/**
 * This is a stub Camel From in JSON format.
 * It is used to test the Canvas component.
 */
export const camelFromJson: { from: FromDefinition } = {
  from: {
    uri: 'timer:tutorial',
    steps: [
      {
        to: {
          uri: 'direct:my-route',
        },
      },
    ],
  },
};

export const camelFrom = new CamelRouteVisualEntity({ from: camelFromJson.from });
