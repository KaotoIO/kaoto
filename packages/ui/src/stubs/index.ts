export * from './camel-route';
export * from './kamelet-binding-route';
export * from './kamelet-route';
export * from './pipe';
