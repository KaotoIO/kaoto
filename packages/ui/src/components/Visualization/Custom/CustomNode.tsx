import { Tooltip } from '@patternfly/react-core';
import { ArrowDownIcon, ArrowUpIcon, BanIcon, CodeBranchIcon, PlusIcon } from '@patternfly/react-icons';
import {
  ContextMenuSeparator,
  Decorator,
  DefaultNode,
  ElementModel,
  GraphElement,
  Node,
  NodeStatus,
  WithSelectionProps,
  observer,
  withContextMenu,
  withSelection,
} from '@patternfly/react-topology';
import clsx from 'clsx';
import { FunctionComponent, ReactElement, useContext } from 'react';
import { AddStepMode } from '../../../models/visualization/base-visual-entity';
import { CanvasDefaults } from '../Canvas/canvas.defaults';
import { CanvasNode } from '../Canvas/canvas.models';
import './CustomNode.scss';
import { ItemAddStep } from './ItemAddStep';
import { ItemDeleteGroup } from './ItemDeleteGroup';
import { ItemDeleteStep } from './ItemDeleteStep';
import { ItemDisableStep } from './ItemDisableStep';
import { ItemInsertStep } from './ItemInsertStep';
import { ItemReplaceStep } from './ItemReplaceStep';
import { doTruncateLabel } from '../../../utils/truncate-label';
import { SettingsContext } from '../../../providers';

interface CustomNodeProps extends WithSelectionProps {
  element: Node<CanvasNode, CanvasNode['data']>;
}
const noopFn = () => {};

const CustomNode: FunctionComponent<CustomNodeProps> = observer(({ element, ...rest }) => {
  const vizNode = element.getData()?.vizNode;
  const settingsAdapter = useContext(SettingsContext);
  const label = vizNode?.getNodeLabel(settingsAdapter.getSettings().nodeLabel);
  const isDisabled = !!vizNode?.getComponentSchema()?.definition?.disabled;
  const tooltipContent = vizNode?.getTooltipContent();
  const statusDecoratorTooltip = vizNode?.getNodeValidationText();
  const nodeStatus = !statusDecoratorTooltip || isDisabled ? NodeStatus.default : NodeStatus.warning;

  return (
    <DefaultNode
      {...rest}
      element={element}
      label={doTruncateLabel(label)}
      labelClassName={clsx('custom-node__label', { 'custom-node__label--disabled': isDisabled })}
      showStatusDecorator={!isDisabled}
      statusDecoratorTooltip={statusDecoratorTooltip}
      nodeStatus={nodeStatus}
      onStatusDecoratorClick={noopFn}
    >
      <g
        className="custom-node"
        data-testid={`custom-node__${vizNode?.id}`}
        data-nodelabel={label}
        data-disabled={isDisabled}
      >
        <foreignObject
          x="0"
          y="0"
          width={CanvasDefaults.DEFAULT_NODE_DIAMETER}
          height={CanvasDefaults.DEFAULT_NODE_DIAMETER}
        >
          <Tooltip content={tooltipContent}>
            <div className="custom-node__image">
              <img src={vizNode?.data.icon} />
            </div>
          </Tooltip>
        </foreignObject>

        {isDisabled && (
          <Decorator
            radius={12}
            x={CanvasDefaults.DEFAULT_NODE_DIAMETER}
            y={0}
            icon={<BanIcon className="custom-node--disabled" />}
            showBackground
          />
        )}
      </g>
    </DefaultNode>
  );
});

export const CustomNodeWithSelection: typeof DefaultNode = withSelection()(
  withContextMenu((element: GraphElement<ElementModel, CanvasNode['data']>) => {
    const items: ReactElement[] = [];
    const vizNode = element.getData()?.vizNode;
    if (!vizNode) return items;

    const nodeInteractions = vizNode.getNodeInteraction();

    if (nodeInteractions.canHavePreviousStep) {
      items.push(
        <ItemAddStep
          key="context-menu-item-prepend"
          data-testid="context-menu-item-prepend"
          mode={AddStepMode.PrependStep}
          vizNode={vizNode}
        >
          <ArrowUpIcon /> Prepend
        </ItemAddStep>,
      );
    }
    if (nodeInteractions.canHaveNextStep) {
      items.push(
        <ItemAddStep
          key="context-menu-item-append"
          data-testid="context-menu-item-append"
          mode={AddStepMode.AppendStep}
          vizNode={vizNode}
        >
          <ArrowDownIcon /> Append
        </ItemAddStep>,
      );
    }
    if (nodeInteractions.canHavePreviousStep || nodeInteractions.canHaveNextStep) {
      items.push(<ContextMenuSeparator key="context-menu-separator-add" />);
    }

    if (nodeInteractions.canHaveChildren) {
      items.push(
        <ItemInsertStep
          key="context-menu-item-insert"
          data-testid="context-menu-item-insert"
          mode={AddStepMode.InsertChildStep}
          vizNode={vizNode}
        >
          <PlusIcon /> Add step
        </ItemInsertStep>,
      );
    }
    if (nodeInteractions.canHaveSpecialChildren) {
      items.push(
        <ItemInsertStep
          key="context-menu-item-insert-special"
          data-testid="context-menu-item-insert-special"
          mode={AddStepMode.InsertSpecialChildStep}
          vizNode={vizNode}
        >
          <CodeBranchIcon /> Add branch
        </ItemInsertStep>,
      );
    }
    if (nodeInteractions.canHaveChildren || nodeInteractions.canHaveSpecialChildren) {
      items.push(<ContextMenuSeparator key="context-menu-separator-insert" />);
    }

    if (nodeInteractions.canBeDisabled) {
      items.push(
        <ItemDisableStep key="context-menu-item-disable" data-testid="context-menu-item-disable" vizNode={vizNode} />,
      );
    }
    if (nodeInteractions.canReplaceStep) {
      items.push(
        <ItemReplaceStep key="context-menu-item-replace" data-testid="context-menu-item-replace" vizNode={vizNode} />,
      );
    }
    if (nodeInteractions.canBeDisabled || nodeInteractions.canReplaceStep) {
      items.push(<ContextMenuSeparator key="context-menu-separator-replace" />);
    }

    if (nodeInteractions.canRemoveStep) {
      const childrenNodes = vizNode.getChildren();
      const shouldConfirmBeforeDeletion = childrenNodes !== undefined && childrenNodes.length > 0;
      items.push(
        <ItemDeleteStep
          key="context-menu-item-delete"
          data-testid="context-menu-item-delete"
          vizNode={vizNode}
          loadModal={shouldConfirmBeforeDeletion}
        />,
      );
    }
    if (nodeInteractions.canRemoveFlow) {
      items.push(
        <ItemDeleteGroup
          key="context-menu-container-remove"
          data-testid="context-menu-container-remove"
          vizNode={vizNode}
        />,
      );
    }

    return items;
  })(CustomNode as typeof DefaultNode),
) as typeof DefaultNode;
