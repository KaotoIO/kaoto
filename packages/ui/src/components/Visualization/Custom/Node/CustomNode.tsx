import { Tooltip } from '@patternfly/react-core';
import { BanIcon } from '@patternfly/react-icons';
import {
  Decorator,
  DefaultNode,
  Node,
  NodeStatus,
  ScaleDetailsLevel,
  WithSelectionProps,
  observer,
  useSelection,
  withContextMenu,
  withSelection,
} from '@patternfly/react-topology';
import clsx from 'clsx';
import { FunctionComponent, useContext } from 'react';
import { SettingsContext } from '../../../../providers';
import { doTruncateLabel } from '../../../../utils/truncate-label';
import { CanvasDefaults } from '../../Canvas/canvas.defaults';
import { CanvasNode } from '../../Canvas/canvas.models';
import './CustomNode.scss';
import { NodeContextMenuFn } from '../ContextMenu/NodeContextMenu';

interface CustomNodeProps extends WithSelectionProps {
  element: Node<CanvasNode, CanvasNode['data']>;
}
const noopFn = () => {};

const CustomNode: FunctionComponent<CustomNodeProps> = observer(({ element, ...rest }) => {
  const vizNode = element.getData()?.vizNode;
  const settingsAdapter = useContext(SettingsContext);
  const label = vizNode?.getNodeLabel(settingsAdapter.getSettings().nodeLabel);
  const isDisabled = !!vizNode?.getComponentSchema()?.definition?.disabled;
  const tooltipContent = vizNode?.getTooltipContent();
  const statusDecoratorTooltip = vizNode?.getNodeValidationText();
  const nodeStatus = !statusDecoratorTooltip || isDisabled ? NodeStatus.default : NodeStatus.warning;
  const detailsLevel = element.getGraph().getDetailsLevel();
  const [selected] = useSelection();

  return (
    <DefaultNode
      {...rest}
      element={element}
      label={selected ? label : doTruncateLabel(label)}
      scaleLabel={detailsLevel !== ScaleDetailsLevel.low}
      labelClassName={clsx('custom-node__label', {
        'custom-node__label--disabled': isDisabled,
        'custom-node__label--selected': selected,
      })}
      showStatusDecorator={!isDisabled}
      statusDecoratorTooltip={statusDecoratorTooltip}
      nodeStatus={nodeStatus}
      onStatusDecoratorClick={noopFn}
    >
      <g
        className="custom-node"
        data-testid={`custom-node__${vizNode?.id}`}
        data-nodelabel={label}
        data-disabled={isDisabled}
      >
        <foreignObject
          x="0"
          y="0"
          width={CanvasDefaults.DEFAULT_NODE_DIAMETER}
          height={CanvasDefaults.DEFAULT_NODE_DIAMETER}
        >
          <Tooltip content={tooltipContent}>
            <div className="custom-node__image">
              <img src={vizNode?.data.icon} />
            </div>
          </Tooltip>
        </foreignObject>

        {isDisabled && (
          <Decorator
            radius={12}
            x={CanvasDefaults.DEFAULT_NODE_DIAMETER}
            y={0}
            icon={<BanIcon className="custom-node--disabled" />}
            showBackground
          />
        )}
      </g>
    </DefaultNode>
  );
});

export const CustomNodeWithSelection: typeof DefaultNode = withSelection()(
  withContextMenu(NodeContextMenuFn)(CustomNode as typeof DefaultNode),
) as typeof DefaultNode;
