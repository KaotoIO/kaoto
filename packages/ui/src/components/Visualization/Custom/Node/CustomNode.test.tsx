import { BaseEdge, BaseGraph, BaseNode, ElementContext, VisualizationProvider } from '@patternfly/react-topology';
import { act, render } from '@testing-library/react';
import { TestProvidersWrapper } from '../../../../stubs';
import { ControllerService } from '../../Canvas/controller.service';
import { CustomNodeObserver } from './CustomNode';

describe('CustomNode', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should throw an error if not used on Node elements', () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    const edgeElement = new BaseEdge();

    expect(() => {
      act(() => {
        render(<CustomNodeObserver element={edgeElement} />);
      });
    }).toThrow('CustomNode must be used only on Node elements');
  });

  it('should render without error', () => {
    const parentElement = new BaseGraph();
    const element = new BaseNode();
    const controller = ControllerService.createController();
    parentElement.setController(controller);
    element.setController(controller);
    element.setParent(parentElement);

    const { Provider } = TestProvidersWrapper();

    const wrapper = render(
      <Provider>
        <VisualizationProvider controller={controller}>
          <ElementContext.Provider value={element}>
            <CustomNodeObserver element={element} />
          </ElementContext.Provider>
        </VisualizationProvider>
      </Provider>,
    );

    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
