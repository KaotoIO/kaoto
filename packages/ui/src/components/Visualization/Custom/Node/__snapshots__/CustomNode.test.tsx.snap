// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomNode should render without error 1`] = `<DocumentFragment />`;
